### DISCUSSION TITLE
Recommended replacement for ESLint `traverser` class?

### DISCUSSION THREAD

**Discussion by bmish** [eslint/eslint#15154] [Q&A] [answered]
ESLint's [lib/shared/traverser.js](https://github.com/eslint/eslint/blob/master/lib/shared/traverser.js) class is no longer accessible as of ESLint v8. Due to ESM adoption, we can no longer import it like this:

```js
const traverser = require('eslint/lib/shared/traverser');
```

I was using this class in a number of plugin rules to conveniently traverse the AST or parts of the AST at various points outside of the standard, one-time traversal that ESLint performs of each AST. For example, here's a neat little function for finding if there's a matching node inside another node using the traverser:

```js
function hasMatchingNode(node, matcher) {
  let foundMatch = false;
  new Traverser().traverse(node, {
    enter(child) {
      if (!foundMatch && matcher(child)) {
        foundMatch = true;
      }
    },
  });
  return foundMatch;
}
```

I'm wondering if there is a recommended replacement for this traversal functionality (either from ESLint or a popular/well-maintained third-party package).

**Comment by bmish** [comment, marked as answer]
It looks like [estraverse](https://github.com/estools/estraverse) is a pretty good, one-to-one replacement.

**Reply by nzakas** [reply]
Yup! Traverser does the same thing, just in a way where we can manipulate the traversal path for things like code path analysis.
